### ISSUE TITLE
Only allow user A to award user B one delta for the SAME comment

### ISSUE THREAD

**Issue by Snorrrlax** [MystK/delta-bot-three#50]
I put "same" in caps because it's important this isn't confused with other scenarios where user A *is* allowed to award user B a second delta (i.e for a separate comment anywhere else in the thread).

[Here](https://www.reddit.com/r/changemyview/comments/560137/cmv_people_should_be_punished_for_attempted/d8f8pc9) you can see /u/lak16 awarded a delta twice with a duplicate comment. DeltaBot should reject it and reply with something along the lines of "You have already awarded a delta for this comment".

**Comment by roastchicken** [comment]
Looks like this is already checked for. Unless someone manually removed /u/lak16's second delta, /u/tylerthehun only recieved one delta from them, as shown in [/u/tylerthehun's wiki page](https://www.reddit.com/r/changemyview/wiki/user/tylerthehun).

**Comment by Snorrrlax** [comment]
Ah okay, DeltaBot doesn't add a second delta but gives a duplicate response. This should still be fixed to say "you've already awarded a delta for this comment".

**Comment by roastchicken** [comment]
@Snorrrlax could I be assigned to this? I'm going to try to work on it today.

**Comment by MystK** [comment]
Probably leaving a comment like this is more than adequate so when someone else opens the ticket they can see.

**Comment by Snorrrlax** [comment]
@roastchicken Thanks! In that case I've added the "in progress" label to this issue, as per my updates to #100

**Comment by Snorrrlax** [comment]
A couple more thoughts on this issue: 

It would be a rare case, but this shouldn't continue indefinitely - it would open up DeltaBot for trolling. In other words, if user A awards user B a second delta for the *same* comment, then DeltaBot should leave a comment as mentioned in this issue. But if user A was to award user B a third, fourth, fifth (and so on), then DeltaBot shouldn't respond to it - but instead report it.

You may have already considered this, but it DeltaBot should check if user A's first delta to user B's comment was accepted before rejecting the second one. User A may have had their first decline due to character count etc, and decided to make a new comment with the expanded explanation.

**Comment by roastchicken** [comment]
What would the bot do to report the comment?

Also, I have finals today and tomorrow so I don't think I'll get to working on this until Wednesday.

**Comment by Snorrrlax** [comment]
Perhaps @mystk could answer this, I just assumed that since DeltaBot can "click" reply on a comment it could also "click" report (with a report reason). If it turns out that adding a report function requires a lot more code than I'm assuming, then feel free to leave that out and just have DeltaBot ignore instead.

**Comment by roastchicken** [comment]
Okay, I see. I though the report button reported it to the reddit admins, but I just checked it out and I see that I'm wrong about that. I'm pretty sure a new function will have to be added to the Reddit API that DeltaBot is using in order to report a post, and I'm not sure how hard that would be. I'll just ignore subsequent delta attempts after the first two, and once I have more time I'll look into adding the report functionality.

**Comment by Snorrrlax** [comment]
Hi @roastchicken, any update on this?

**Comment by roastchicken** [comment]
None yet :(

I've been meaning to work on this but just haven't found the willpower. I'm going to set up a dev environment for DeltaBot again because I had to use my other VPS for something else and I was running low on disk space. Hopefully starting on this will make it easier to just get it over with, but who knows.

**Comment by Snorrrlax** [comment]
Okay, fingers crossed. Perhaps a WIP pull request like wyantb submitted would help? 

Also, is this issue a duplicate of #28?

**Comment by roastchicken** [comment]
I don't see any check for a user awarding someone a delta twice in verifyThenAward, so I'm unsure where to implement a comment. Does DeltaBot actually prevent double issuing of deltas, or was it just a fluke? @MystK might be able to shed some light on this, as he knows the code base better than anyone.

**Comment by roastchicken** [comment]
Also, @Snorrrlax after taking a look at #28 this doesn't seem to be a duplicate of it. #28 is an issue regarding DeltaBot commenting twice on _one_ delta-awarding comment, whereas this issue is in regards to DeltaBot commenting twice on _two_ delta-awarding comments made by the same user.

**Comment by MystK** [comment]
Hey there, I don't think there is a check if DB has replied when awarding a delta, but there is some code in another section to check if DB replied. We could modularize that code and use it in both cases. 

I believe the verify then award section the the correct section.

**Comment by Snorrrlax** [comment]
@roastchicken Any update on this?

**Comment by Snorrrlax** [comment]
@roastchicken do you think you'll have time for this soon?

**Comment by Snorrrlax** [comment]
@roastchicken Hi, I just want to check if you still plan to work on this?

**Comment by roastchicken** [comment]
@Snorrrlax sorry, I've been busy with stuff. I don't know if/where DeltaBot checks to see if it has already replied, and thus cannot progress on this issue. If @MystK or someone else knows where that code is located, I could take a look at it and attempt to modify it to be used for the new use case.

**Comment by Snorrrlax** [comment]
@roastchicken No worries, hopefully @MystK can clear things up for you soon and we can cross this off the list.

**Comment by MystK** [comment]
What snorrrlax says here and in this [mod comment](https://www.reddit.com/r/changemyview/comments/5z8sj5/cmv_redditcom_is_an_awesome_website/dewigq7/) is conflicting.

[For how DB checks for a reply, see line 126 of the utils.js file.](https://github.com/MystK/delta-bot-three/blob/master/src/utils.js#L126)

**Comment by Snorrrlax** [comment]
Ah, there's been some miscommunication between the mods that I'll sort out. Please use what I've been saying here unless I come back with a change after discussing with that mod.

**Comment by roastchicken** [comment]
I haven't checked out that line yet, but this is great timing. I'm just about to start Spring break so I should hopefully have plenty of time to work on this.

**Comment by Snorrrlax** [comment]
Any update on this?

**Comment by Snorrrlax** [comment]
@roastchicken are you still interested in helping with this issue?

**Comment by roastchicken** [comment]
@Snorrrlax yes! I really want to finish this once and for all. I'll take a look at it again tomorrow. Thanks for the reminder.

**Comment by roastchicken** [comment]
I've looked at [getDeltaBotReply](https://github.com/MystK/delta-bot-three/blob/master/src/utils.js#L126-L144), but my lack of experience with JavaScript is making it difficult for me to understand. @MystK could you give me an example of what the replies parameter might look like? And also, does this function always return a boolean?

**Comment by MystK** [comment]
Hey @roastchicken, the easiest way to work on this ticket would be to get the bot running on your machine and debugging the code to see what goes on. It's pretty difficult if not impossible unless you do this step since you need to test it anyways. Did you have it set up? Did you need help?

**Comment by roastchicken** [comment]
@MystK 
No, I do not yet have a running test environment setup. At least, I don't think so. I had one setup on one of my VPSes, but I don't think it's there anymore. I'll have to try to be sure, but I don't think I'll have any issues setting it up again.

By debugging, what do you mean exactly? Just `console.log`ing variables to see their values? Or is there a better way to debug JavaScript/node?

The reason I asked you about getDeltaBotReply instead of setting up an environment and using `console.log` to get example values was mainly laziness. From what I understand you know the codebase the best, any maybe you knew off the top of your head (or could figure out quickly by looking at the function) what the function parameter and return value were. Maybe that's not true; on all my projects I usually know what values variables typically have, but I've also never worked on something close to the scale of DeltaBot.

I'll work on getting a test environment setup.
